File: basic_dsa_four_hundred_fifty_three.java
/**
 * Basic DSA Problems - Minimum Moves to Equal Array Elements (LeetCode 453)
 * Each move: increment n-1 elements by 1. Min moves to make all equal.
 */
public class basic_dsa_four_hundred_fifty_three {

    // Problem: [1,2,3] -> 3 moves. Increment n-1 = decrement 1.

    /**
     * Equivalent to decrementing 1 element per move. Ans = sum - min*n.
     * Time O(n), Space O(1)
     */
    public static int minMoves(int[] nums) {
        int min = nums[0], sum = 0;
        for (int x : nums) {
            if (x < min) min = x;
            sum += x;
        }
        return sum - min * nums.length;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Minimum Moves to Equal Array Elements");
        System.out.println("=".repeat(60));

        System.out.println("[1,2,3] -> " + minMoves(new int[]{1,2,3}));
        System.out.println("[1,1,1] -> " + minMoves(new int[]{1,1,1}));
        System.out.println("[1,2,3,4] -> " + minMoves(new int[]{1,2,3,4}));
    }
}

File: basic_dsa_four_hundred_fifty_three.cpp
/**
 * Basic DSA - Sum of Unique Elements (LeetCode 1748)
 * Sum of numbers that appear exactly once in nums.
 */
#include <iostream>
#include <unordered_map>
#include <vector>

int sumOfUnique(const std::vector<int>& nums) {
    std::unordered_map<int, int> cnt;
    for (int x : nums)
        ++cnt[x];
    int s = 0;
    for (const auto& p : cnt) {
        if (p.second == 1)
            s += p.first;
    }
    return s;
}

void printVec(const std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i)
            std::cout << ", ";
        std::cout << v[i];
    }
    std::cout << "]";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Sum of Unique Elements\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {1, 2, 3, 2};
    printVec(a1);
    std::cout << " -> " << sumOfUnique(a1) << "\n";

    std::vector<int> a2 = {1, 1, 1, 1, 1};
    printVec(a2);
    std::cout << " -> " << sumOfUnique(a2) << "\n";

    std::vector<int> a3 = {1, 2, 3, 4, 5};
    printVec(a3);
    std::cout << " -> " << sumOfUnique(a3) << "\n";
}
